src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -225,24 +245,14 @@ private int LookupAggregatorStore(string[] tagKeys, object[] tagValues, int leng
                             return -1;
                         }
 
-                        // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
-                        if (seqKey == null)
-                        {
-                            seqKey = new string[length];
-                            tagKeys.CopyTo(seqKey, 0);
-                        }
-
-                        var seqVal = new object[length];
-                        tagValues.CopyTo(seqVal, 0);
-
                         ref var metricPoint = ref this.metricPoints[aggregatorIndex];
                         var dt = DateTimeOffset.UtcNow;
-                        metricPoint = new MetricPoint(this.aggType, dt, seqKey, seqVal, this.histogramBounds);
+                        metricPoint = new MetricPoint(this.aggType, dt, sortedTags.Keys, sortedTags.Values, this.histogramBounds);
 
                         // Add to dictionary *after* initializing MetricPoint
                         // as other threads can start writing to the
                         // MetricPoint, if dictionary entry found.
-                        value2metrics.TryAdd(seqVal, aggregatorIndex);
+                        this.tagsToMetricPointIndexDictionary.TryAdd(sortedTags, aggregatorIndex);
```

Comment:
to confirm the approach here : with this PR, we'll now have to repeatedly store the key and value for every entry, as opposed to the existing code, which stored keys only once.

This would lead to memory wastage storing duplicates entries, but if the benefits outweigh, then we can move forward.

Reply:
IncomingTags
Lookup IncomingTags->Index
Success: use index
Fail: 
Lookup SortedIncomingTags->Index
             Success: use index
             Failure:
             Insert IncomingTags->Index, SortedIncomingTags->Index to dictionary.


^ Is the above better? didn't think fully through.. but it avoids memory explosion. at max 2 entries. and there is only a single dictionary

Reply:
This could work.

Reply:
I have implemented the above suggestion of using just one dictionary. With this approach, we store at most two entries for any given MetricPoint. (sorted + an optional unsorted combination).

Here are the Benchmark results:

// * Summary *

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
Intel Core i7-9700 CPU 3.00GHz, 1 CPU, 8 logical and 8 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT


|                    Method | AggregationTemporality |      Mean |    Error |   StdDev | Allocated |
|-------------------------- |----------------------- |----------:|---------:|---------:|----------:|
|            CounterHotPath |             Cumulative |  15.78 ns | 0.024 ns | 0.022 ns |         - |
| CounterWith1LabelsHotPath |             Cumulative |  56.90 ns | 0.043 ns | 0.036 ns |         - |
| CounterWith3LabelsHotPath |             Cumulative | 144.11 ns | 0.492 ns | 0.461 ns |         - |
| CounterWith5LabelsHotPath |             Cumulative | 227.84 ns | 0.692 ns | 0.578 ns |         - |
| CounterWith6LabelsHotPath |             Cumulative | 260.65 ns | 0.640 ns | 0.599 ns |         - |
| CounterWith7LabelsHotPath |             Cumulative | 289.10 ns | 1.119 ns | 0.934 ns |         - |
|            CounterHotPath |                  Delta |  15.98 ns | 0.219 ns | 0.194 ns |         - |
| CounterWith1LabelsHotPath |                  Delta |  62.86 ns | 0.157 ns | 0.147 ns |         - |
| CounterWith3LabelsHotPath |                  Delta | 141.15 ns | 0.415 ns | 0.368 ns |         - |
| CounterWith5LabelsHotPath |                  Delta | 224.20 ns | 1.083 ns | 0.960 ns |         - |
| CounterWith6LabelsHotPath |                  Delta | 257.57 ns | 0.670 ns | 0.594 ns |         - |
| CounterWith7LabelsHotPath |                  Delta | 286.83 ns | 1.061 ns | 0.941 ns |         - |

Reply:
Here the Stress Test results with the new changes to the PR. Loops per second has increased to **~29M**.

![image](https://user-images.githubusercontent.com/66651184/151463915-0f3d31bc-b1db-4527-a50a-6c07bb2a614a.png)